json.go

```diff
@@ -0,0 +1,809 @@
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+	"unsafe"
+)
+
+type JSONOption uint8
+
+var ErrInvalidJSON = errors.New("Invalid JSON Document")
+var ErrInvalidType = errors.New("Invalid JSON Scalar Type")
+
+const (
+	JSONOptDefault        = JSONOption(C.DPI_JSON_OPT_DEFAULT)
+	JSONOptNumberAsString = JSONOption(C.DPI_JSON_OPT_NUMBER_AS_STRING)
+	JSONOptDateAsDouble   = JSONOption(C.DPI_JSON_OPT_DATE_AS_DOUBLE)
+)
+
+// Represents JSON string format, if its in std, ORACLE extended types,
+// BSON extended types, etc
+type JSONStringFlags uint
+
+const (
+	JSONFormatExtnTypes       JSONStringFlags = C.DPI_JSON_USE_EXTENSION_TYPES
+	JSONFormatBSONTypes                       = C.DPI_JSON_BSON_TYPE_PATTERNS
+	JSONFormatBSONTypePattern                 = C.DPI_JSON_USE_BSON_TYPES
+)
+
+// Encapsulates JSON formatted string.
+// it could be reader/writer for streams.
+type JSONString struct {
+	Flags JSONStringFlags // standard , extended types for OSON, BSON
+	Value string          // JSON input
+}
+
+// JSONValue is the interface implemented by structs encapsulating
+// various JSON inputs(map, array, struct(?),scalar values like int, string ..).
+// JSON, JSONObject, JSONArray and JSONScalar implement it.
+type JSONValue interface {
+	// JSONType is Oracle type stored. Its not used now.
+	getOracleJSONType() C.dpiOracleTypeNum
+}
+
+// JSONScalar holds the JSON data to/from Oracle.
+// It includes all scalar values such as int, float, bool, time.Time,
+// time.Duration, byte[] , map and arrays.
+type JSONScalar struct {
+	dpiJsonNode *C.dpiJsonNode
+	isMemOwned  bool // true indicates associated C memory needs to be freed.
+}
+
+// Returns the Go type not the DPI native/oracle type
+func (j JSONScalar) GetValue() (val interface{}, err error) {
+	var d Data
+	jsonNodeToData(&d, j.dpiJsonNode)
+	if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_NUMBER {
+		val, err = getJSONScalarDouble(d)
+	} else if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_VARCHAR {
+		val, err = getJSONScalarString(d)
+	} else {
+		val = d.Get()
+		if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_JSON_OBJECT {
+			jobj := val.(JSONObject)
+			val, err = jobj.GetValue()
+		} else if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_JSON_ARRAY {
+			jarr := val.(JSONArray)
+			val, err = jarr.GetValue()
+		} else {
+			err = nil
+		}
+	}
+	return
+}
+
+func (j JSONScalar) getOracleJSONType() C.dpiOracleTypeNum {
+	return j.dpiJsonNode.oracleTypeNum
+}
+
+func (js *JSONScalar) Close() error {
+	if js.isMemOwned {
+		C.godror_dpiJsonfreeMem(js.dpiJsonNode)
+	}
+	return nil
+}
+
+// JSON holds the JSON data to/from Oracle.
+// It is like a root node in JSON tree.
+type JSON struct {
+	dpiJson       *C.dpiJson
+	isMemOwned    bool // true indicates associated C memory needs to be freed.
+	oracleTypeNum C.dpiOracleTypeNum
+}
+
+// During Fetch the JSON struct is created.
+// Explicit creation is not supported.
+func createJSON(js *C.dpiJson, mutable bool) (JSON, error) {
```

Comment:
Why the error return when it never returns error?

Reply:
Fixed it.